compiler/src/iree/compiler/Dialect/HAL/IR/HALOps.td

```diff
@@ -685,6 +685,116 @@ def HAL_BufferAssertOp : HAL_Op<"buffer.assert"> {
   // (such as when we create it ourselves earlier on) or we've already asserted.
 }
 
+def HAL_BufferAllocationPreserveOp : HAL_Op<"buffer.allocation.preserve"> {
```

Comment:
Do we want to have some side-effecting to this? I know today MLIR transformations (like CSE or DCE) treat ops with no return types in a special way. But this op has no results, would it require some re-ordering constraints? I dont know how this is represented in IR, but an owner should not access `%arg0` before taking co-ownership?

Reply:
`Arg<T, [MemWrite]>:$argument` for example, works to specify the effects on a particular input.

Reply:
Adding them here but filed #20328 to add them elsewhere.